### PR TITLE
[FEATURE] Dnnl sum primitive path

### PR DESCRIPTION
**Pull request by Kacper-Pietkun**
Added path to DNNLSumFwd in broadcast binary add, when inputs' shapes are the same. Overall speedup is not noticeable, however, it is significant for some edge cases.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

File: src/operator/tensor/elemwise_binary_broadcast_op_basic.cc [hunk 1]
```diff
@@ -38,31 +39,39 @@ void DNNLBinaryOpForward(const nnvm::NodeAttrs& attrs,
...
-    new_inputs  = {inputs[0].Reshape(new_lshape), inputs[1].Reshape(new_rshape)};
-    new_outputs = {outputs[0].Reshape(new_oshape)};
-  } else if (inputs[0].shape().Size() == 1 && inputs[1].shape().Size() == 1) {
-    // BinaryBroadcastShapeCompact function doesn't reshape tensors of size (1,1,...,1)
-    // into shape (1). It is mandatory for oneDNN primitive to have this reshape done.
-    mxnet::TShape one_shape = mxnet::TShape(1, 1);
-    new_inputs              = {inputs[0].Reshape(one_shape), inputs[1].Reshape(one_shape)};
-    new_outputs             = {outputs[0].Reshape(one_shape)};
+  // We can use more efficient sum kernel when there is no broadcast - when shapes are the same
+  const bool same_shape = (inputs[0].shape() == inputs[1].shape());
```
[the remark this answers is not in the document]
Comment (bartekkuncer) [reply src/operator/tensor/elemwise_binary_broadcast_op_basic.cc:43]: Maybe put these shapes in variables as they are referred to multiple times?

File: src/operator/tensor/elemwise_binary_broadcast_op_basic.cc [hunk 2]
```diff
@@ -38,31 +39,39 @@ void DNNLBinaryOpForward(const nnvm::NodeAttrs& attrs,
...
+                                                &new_lshape,
+                                                &new_rshape,
+                                                &new_oshape);
+    std::vector<NDArray> new_inputs;
+    std::vector<NDArray> new_outputs;
+    if (ndim_diff) {
+      new_inputs  = {inputs[0].Reshape(new_lshape), inputs[1].Reshape(new_rshape)};
+      new_outputs = {outputs[0].Reshape(new_oshape)};
+    } else if (inputs[0].shape().Size() == 1 && inputs[1].shape().Size() == 1) {
+      // BinaryBroadcastShapeCompact function doesn't reshape tensors of size (1,1,...,1)
```
Comment (bartekkuncer) [inline src/operator/tensor/elemwise_binary_broadcast_op_basic.cc:62]:
```suggestion
      // BinaryBroadcastShapeCompact function does not reshaping tensors of size (1,1,...,1)
```

File: src/operator/tensor/elemwise_binary_broadcast_op_basic.cc [hunk 3]
```diff
@@ -38,31 +39,39 @@ void DNNLBinaryOpForward(const nnvm::NodeAttrs& attrs,
...
+      // into shape (1). It is mandatory for oneDNN primitive to have this reshape done.
```
Comment (bartekkuncer) [inline src/operator/tensor/elemwise_binary_broadcast_op_basic.cc:63]: So BinaryBroadcastShapeCompact does not support this reshape or oneDNN?

File: tests/python/unittest/test_operator.py [hunk 4]
```diff
@@ -9414,6 +9414,24 @@ def test_elementwise_ops_on_misaligned_input():
...
     assert a[3].asscalar() == 4.0
 
+
+@pytest.mark.parametrize('dtype', ['float16', 'float32', 'float64'])
+@pytest.mark.parametrize('ndim', [1, 2, 3, 4, 5])
+@pytest.mark.parametrize('max_dim_size', [1, 2, 3, 4, 5])
+def test_broadcast_ops_on_input_with_the_same_shape(dtype, ndim, max_dim_size):
+    shape = list(rand_shape_nd(ndim, dim=max_dim_size))
+    shape_size = np.product(shape)
+    a = np.random.uniform(low=-100, high=100, size=5000)
```
Comment (bartekkuncer) [inline tests/python/unittest/test_operator.py:9424]: Please change it to 3125 and add simple comment that it is 5^5/5**5.

**Review by bartekkuncer — CHANGES_REQUESTED** [review]

File: src/operator/tensor/elemwise_binary_broadcast_op_basic.cc (hunk 3 above)
Comment (Kacper-Pietkun) [reply src/operator/tensor/elemwise_binary_broadcast_op_basic.cc:63]: BinaryBroadcastShapeCompact  does not support reshape of tensors of size (1,1,...,1), but we need to reshape such tensors into shape (1) for oneDNN to work properly

File: tests/python/unittest/test_operator.py (hunk 4 above)
Comment (Kacper-Pietkun) [reply tests/python/unittest/test_operator.py]: Changed size to use shape.

**Review by bartekkuncer — APPROVED** [review]

**Comment by Kacper-Pietkun** [comment]
@mxnet-bot run [centos-gpu, unix-gpu]

**Comment by Kacper-Pietkun** [comment]
@mxnet-bot run ci [centos-gpu, unix-gpu]

**Comment by Kacper-Pietkun** [comment]
@mxnet-bot run ci [unix-gpu]

**Review by bgawrych — APPROVED** [review]

**Review by bartekkuncer — APPROVED** [review]

**Pull request merged**